### PR TITLE
Add schema logic in producer and consumer

### LINKED ISSUE #344: Support Schema Registry feature in Go Client
This is the Schema Registry design doc For Go Client: https://gist.github.com/wolfstudy/e0c9b349cd405715a9e04dd8658510bc

- [x] Basic classes on supporting schema (Primitive, PROTO, JSON, AVRO) #368
- [x] Support schema on Producer @wolfstudy #368
- [x] Support schema on Consumer @wolfstudy #368
- [ ] Add the schema version support - Multiple schema on a single topic (*) 
- [ ] Support schema on Reader
- [ ] Advanced Schema
   - [ ] KeyValue Schema
   - [ ] AutoSchema

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <rxl@apache.org>

Fixes #344 

### Motivation

Type safety is extremely important in any application built around a message bus like Pulsar. Currently, Apache Pulsar supports the function of Schema Registry. And Java, CPP and Python clients already support schema registry related functions, In order to further improve the function of Go Client, we need to support the function of Schema Registry in Go Client.

### Modifications

- Add schema logic in producer
- Add schema logic in consumer
- Package basic tool functions for schema logic

---

### Currente state

- Encapsulates all basic types of decode and encode logic
- Encapsulates the use of all schema types
- A new interface has been added for producer and consumer to be compatible with schema functions

### CODE REVIEW

**Comment by wolfstudy** [comment]
@merlimat @cckellogg PTAL thanks.

File: pulsar/schemaDef_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (cckellogg) [inline pulsar/schemaDef_test.go:1]: rename file schema_test.go. We should use snake_case to be consistent with the rest of the files. Same for other camelCase files

File: pulsar/schemaDef_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+package pulsar
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSchemaDef(t *testing.T) {
+	errSchemaDef := "{\"type\":\"record\",\"name\":\"Example\",\"namespace\":\"test\"," +
```
Comment (cckellogg) [inline pulsar/schemaDef_test.go:27]: Would ` marks make these easier to maintain? We would then not have to scape the " characters

File: pulsar/schema.go [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+type Int16Schema struct {
+	SchemaInfo
+}
+
+func NewInt16Schema(properties map[string]string) *Int16Schema {
+	int16Schema := new(Int16Schema)
+	int16Schema.SchemaInfo.Properties = properties
+	int16Schema.SchemaInfo.Name = "INT16"
+	int16Schema.SchemaInfo.Type = INT16
+	int16Schema.SchemaInfo.Schema = ""
```
Comment (cckellogg) [inline pulsar/schema.go:332]: Why is the schema "" for all primitives?

File: pulsar/primitiveSerDe.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+const (
+	IoMaxSize     = 1024
+	maxBorrowSize = 10
+)
+
+var (
+	littleEndian = binary.LittleEndian
+)
+
+type BinaryFreeList chan []byte
```
Comment (cckellogg) [inline pulsar/primitiveSerDe.go:36]: What's the purpose of this?

File: pulsar/schema.go [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+	"reflect"
+	"unsafe"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/linkedin/goavro"
+)
+
+type SchemaType int
```
Comment (cckellogg) [inline pulsar/schema.go:33]: Should we add a String() method on the enum then for the names we can call JSON.String(), etc... and it's useful for logging and or debugging

File: pulsar/message.go [hunk 6]
```diff
@@ -27,6 +27,9 @@ type ProducerMessage struct {
 	// Payload for the message
 	Payload []byte
 
+	//Value and payload is mutually exclusive, `Value interface{}` for schema message.
+	Value interface{}
```
Comment (cckellogg) [inline pulsar/message.go:31]: Should this be package private?

File: pulsar/schemaDef_test.go (hunk 1 above)
Comment (wolfstudy) [reply pulsar/schemaDef_test.go:1]: Yes, this test case is just to test whether the logic of the schemaDef is correct. When we implement the schema function on the producer and consumer respectively, we will add separate test cases for each schema type. At this time, the file will be named `schema_test.go`.

File: pulsar/message.go (hunk 6 above)
Comment (wolfstudy) [reply pulsar/message.go:31]: Should not be private, this interface is to be exposed to users, for example:

```
err := producer.Send(context.Background(), ProducerMessage{
		Value: &testJson{
			ID:   100,
			Name: "pulsar",
		},
	})
```

File: pulsar/schemaDef_test.go (hunk 2 above)
Comment (wolfstudy) [reply pulsar/schemaDef_test.go:27]: Good ideas will fix it

File: pulsar/schema.go (hunk 5 above)
Comment (wolfstudy) [reply pulsar/schema.go:33]: Maybe we can use SchemaInfo.Name replace it, right?

File: pulsar/schema.go (hunk 3 above)
Comment (wolfstudy) [reply pulsar/schema.go:332]: For primitives types, the schema field should be `""`

File: pulsar/schemaDef_test.go (hunk 1 above)
Comment (cckellogg) [reply pulsar/schemaDef_test.go:1]: we should be consistent with name. all the files are snake_case and I suggest we keep it that way for consistency. schema_def_test.go. Same for the other files in the PR.

**Review by cckellogg — COMMENTED** [review]
I think this PR should be split into 2. First add the structs/interfaces for the schema and then add a follow up PR to implement the schema in the client. This will keep the PRs more self contained and easier to review.

**Comment by skyrocknroll** [comment]
@wolfstudy is there any specific reason we are using v1.0.5 of goavro. According to README `"github.com/linkedin/goavro/v2"` has better support for namespaces and performance improvements. Can we please switch to v2 ?

**Comment by wolfstudy** [comment]
> @wolfstudy is there any specific reason why we are using v1.0.5 of goavro ? According to README `"github.com/linkedin/goavro/v2"` has better support for namespaces and performance improvements. Can we please switch to v2 ?

Cool, I think we can update the `gpavro` to `v2`

**Comment by wolfstudy** [comment]
@skyrocknroll done PTAL again, thanks. 

ping @sijie @massakam If time is convenient, please review it, thanks.

**Review by jiazhai — APPROVED** [review]
lgtm

**Pull request merged**